src/pudl/analysis/mcoe.py

```diff
@@ -499,48 +590,88 @@ def mcoe(
         )
         .pipe(pudl.validate.no_null_cols, df_name="fuel_cost + capacity_factor")
     )
+    mcoe_out = mcoe_out.sort_values(
```

Comment:
Should this MCOE function have these `pudl.helpers.oob_to_nan` calls? Or do they belong somewhere else?

Reply:
I don't think we want to be writing known-to-be-crazy MCOE values into the DB if we can help it.  As this is currently set up, it seems like we'll put unfiltered per-generator capacity factors, fuel costs, and heat rates into the DB in their individual skinny tables, but then filter for realistic values in the combined MCOE table, which is where folks are going to actually access the data I think, which seems like an okay middle ground -- the raw results of our calculations are available, but not highlighted.  We should make it very clear in the descriptions of these assets that they are not intended for use without filtering, and why we expect some of the numbers in them to be totally unreasonable.

Should we parametrize the MCOE constraints in a `context` input to the creation of that asset?

I think @cmgosnell also has thoughts on what we should do here.

Reply:
The fact that we're waiting until the final step to do this filtering, and then only nulling out the individually ridiculous values in each column seems a little bit suspect to me. Since the fuel cost per MWh depends on all of the other values in a row, if any of those values is ridiculous, we should probably null the entire row -- otherwise we can get cases where two numbers that are ridiculous on their own multiply to yield a "reasonable" heat rate / capacity factor / fuel cost value, which we then retain, even though it's based on obviously bogus inputs.

Reply:
okay i think i understand what is all going on here.

I agree that keeping in the baddies in our lil skinny tables and nulling the oobs in this "hey look at me, i'm mcoe" table is a good idea. agreed on the noting the known baddies in the resource defs of the skinny guys.

on the only null-ing the bad field, not the downstream related fields that are maybe deceptively normal looking: what if we included a `related_fields` arg into `oob_to_nan` and pointedly nulled based on known related things? It would feel bad to just null the whole record if one thing was oob.

a question that this all would result in... how do we deal with the validation tests in this new arrangement? rn all of the mcoe validation tests use the mcoe `pudl_out` table with all the min/maxes set to `None` so it keeps all the baddies in there. I **think** a simple enough solution to that would be to set these min/max oob default in this asset, but in the validation tests we could use the skinny all the stinky stuff in there assets as inputs to this function with all `None`s for the min/max.

Reply:
Rather than overloading `oob_to_nan` with more responsibilities maybe we should use it in a new function? Really it should be operating on a single column at a time rather than a whole dataframe but it's an old function.  So the idea would be to say "If you find an out of bounds value in column X, you should null it as well as columns Y and Z (which we know a priori depend on X)?

I agree that we should add validations which expect all of the mess for the skinny and unfiltered MCOE inputs, and then a separate validation which looks at the filtered / cleaned up values in the MCOE table.

With the new outputs is it possible to validate particular prime movers rather than just gross fuel type?  Like can we separately check the heat rates on gas combustion turbines, combined cycle natural gas generators, and coal fired steam turbines?  They should each have pretty distinctive distributions in the clean MCOE outputs.

Reply:
I added a helper function `oob_to_nan_with_dependent_cols` that calls `oob_to_nan` to nullify columns that depend on other columns that have been nulled.

The three columns in the MCOE calculation where `oob_to_nan` is called are:
- `heat_rate_mmbtu_mwh`, used to derive `total_mmbtu`, `fuel_cost_per_mwh`
- `fuel_cost_per_mwh`, no downstream derived columns
- `capacity_factor`, no downstream derived columns